src/combinatorics/ballot.lean

```diff
@@ -0,0 +1,453 @@
+/-
+Copyright (c) 2022 Bhavik Mehta, Kexing Ying. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Kexing Ying
+-/
+import probability.cond_count
+
+/-!
+# Ballot problem
+
+This file proves Theorem 30 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/).
+
+The ballot problem asks, if in an election, candidate A receives `p` votes whereas candidate B
+receives `q` votes where `p > q`, what is the probability that candidate A is strictly ahead
+throughout the count. The probability of this is `(p - q) / (p + q)`.
+
+## Main definitions
+
+* `counted_sequence`: given natural numbers `p` and `q`, `counted_sequence p q` is the set of
+  all lists containing `p` of `1`s and `q` of `-1`s representing the votes of candidate A and B
+  respectively.
+* `stays_positive`: is the set of lists of integers which suffix has positive sum. In particular,
+  the intersection of this set with `counted_sequence` is the set of lists where candidate A is
+  strictly ahead.
+
+## Main result
+
+* `ballot`: the ballot problem.
+
+-/
+
+open set probability_theory measure_theory
+
+/-- The set of nonempty lists of integers which suffix has positive sum. -/
+def stays_positive : set (list ℤ) := {l | ∀ l₂, l₂ ≠ [] → l₂ <:+ l → 0 < l₂.sum}
```

Comment:
This name is too generic. Please either move it to a namespace or move the whole file to `archive/`. BTW, why do you use suffixes instead of prefixes?

Reply:
I used suffix because that is what was done for #5574. @b-mehta is there a reason for using suffix?

Reply:
Primarily because this makes the induction much cleaner: lists in Lean are built up starting from the right, and the natural way to add an element is on the left, so suffixes represent the history of the votes taken. I don't have any particular objection to using the other direction, but I guess the induction will be more difficult

Reply:
I added the namespace after @urkud 's remarks :) I will move the file to `archive` however